File: app/src/main/java/com/mainstreetcode/teammate/model/Prefs.java
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.model;


import android.annotation.SuppressLint;
import android.os.Parcel;
import androidx.annotation.NonNull;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.lang.reflect.Type;

@SuppressLint("ParcelCreator")
public class Prefs implements Model<Prefs> {

    public static Prefs empty() {return new Prefs();}

    private boolean onBoarded;

    private Prefs() {}

    public boolean isOnBoarded() { return onBoarded; }

    public void setOnBoarded(boolean onBoarded) { this.onBoarded = onBoarded; }

    @Override
    public void update(Prefs updated) {
        this.onBoarded = updated.onBoarded;
    }

    @Override
    public int compareTo(@NonNull Prefs o) {
        return 0;
    }

    @Override
    public boolean isEmpty() { return false; }

    @Override
    public String getId() {
        return "1";
    }

    @Override
    public String getImageUrl() {
        return "";
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) { }

    public static class GsonAdapter
            implements
            JsonSerializer<Prefs> {

        private static final String ON_BOARDED = "onBoarded";

        @Override
        public JsonElement serialize(Prefs src, Type typeOfSrc, JsonSerializationContext context) {
            JsonObject serialized = new JsonObject();

            serialized.addProperty(ON_BOARDED, src.onBoarded);

            return serialized;
        }
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/model/Prefs.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.model


import android.annotation.SuppressLint
import android.os.Build.VERSION.SDK_INT
import android.os.Build.VERSION_CODES.Q
import android.os.Parcel
import androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_AUTO_BATTERY
import androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import java.lang.reflect.Type

@SuppressLint("ParcelCreator")
class Prefs private constructor() : Model<Prefs> {

    var isOnBoarded: Boolean = false
    var nightUiMode: Int = if (SDK_INT >= Q) MODE_NIGHT_FOLLOW_SYSTEM else MODE_NIGHT_AUTO_BATTERY

    override val id: String = "1"

    override val isEmpty: Boolean
        get() = false

    override val imageUrl: String
        get() = ""

    override fun update(updated: Prefs) {
        this.isOnBoarded = updated.isOnBoarded
        this.nightUiMode = updated.nightUiMode
    }

    override fun compareTo(other: Prefs): Int = 0

    override fun describeContents(): Int = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = Unit

    class GsonAdapter : JsonSerializer<Prefs> {

        override fun serialize(src: Prefs, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
            val serialized = JsonObject()

            serialized.addProperty(ON_BOARDED, src.isOnBoarded)
            serialized.addProperty(NIGHT_UI_MODE, src.nightUiMode)

            return serialized
        }

        companion object {

            private const val ON_BOARDED = "onBoarded"
            private const val NIGHT_UI_MODE = "nightUiMode"
        }
    }

    companion object {

        fun empty(): Prefs = Prefs()
    }
}
